### DISCUSSION TITLE
Feature Request: graphlet counting

### DISCUSSION THREAD

**Discussion by RobinPetit** [networkit/networkit#850] [Ideas]
Hi, I'd like to use graphlet decomposition of graphs, such as presented in [PGD](https://doi.org/10.1109/ICDM.2015.141) and [ESCAPE](https://doi.org/10.1145/3038912.3052597) but this is not present (yet) in the library.
I would be happy to participate and implement it but I have no idea where in the structure of the lib this should go. Should I create a new directory in `networkit/cpp/` (like `graphlets/` or `topology/`)?

**Comment by avdgrinten** [comment]
For clarification: do you want to count for *all possible subgraphs of size k* how often they occur? Or do you want to count how often *a specific* subgraph of size k occurs?

In any case, a new module sounds reasonable. Maybe `subgraphs/`?

**Comment by RobinPetit** [comment]
>  all possible subgraphs of size k

up to isomorphism, yes. So there are only 4 such graphs of size 3 (2 of which are connected) and 11 such graphs of size 4 (6 of which are connected).

**Comment by RobinPetit** [comment]
Implementation-wise, I see the following possibilities:
1. One class `GraphletsCounter` whose constructor takes an `unsigned int K` parameter and private methods called by the `run` method;
2. One class per parameter K with either the possibility of having classical classes `GraphletsCounter3`, `GraphletsCounter4`, etc. or having a template class with forced specialisations like this:
```C++
template <unsigned int K>
class GraphletsCounter {
public:
    GraphletsCounter() {
        static_assert(
            (K&0x00) > 0,  // ensure this is evaluated at instantiation and not declaration
            "No generic algorithm graphlets counting"
        );
    }
...
};

template<>
class GraphletsCounter<3u> {
...
};

template <>
class GraphletsCounter<4u> {
};
```

I'm not sure the option 1 is really great since while obviously similar in a way, the algorithms are still fairly different and have a different complexity (since in depends on K, the size of the graphlets) so personally I'd go for option 2. What is your opinion on it?

**Reply by avdgrinten** [reply]
Does the algorithm differ for `k = 2` and `k = 3`? If the algorithm is always the same, (1) sounds preferable. If the algorithms are different, we could also have some non-templated `GraphletsCounter` class that dispatches the runtime argument `K` to a templated `template<unsigned int K> GraphletsCounterImpl` implementation (and that fails if `K` is outside of the supported set).

**Reply by RobinPetit** [reply]
For `k = 2`, it is trivial since the only (undirected) graphs of size 2 are `A B` and `A - B` so the number of graphlets are given by `E` and `V(V-1)/2 - E`, therefore the algorithm is `O(1)`. For `k = 3` however, we still need to iterate over all the edges of the graph so we are in `\Omega(E)` (and `O(ED)` for `D` the maximum degree in the graph IIRC).

Your approach with a template implementation class is nice and seems to fit with the general style of the code. I'll do that and submit a PR later then. :+1:
